### PR TITLE
Fix serialization of `pydantic.BaseModel` fields with `pathlib.Path` dict keys

### PR DESCRIPTION
**Pull request by brendanator**
Apply `_simple_default` serialization on dict keys to resolve this exception

```python
Traceback (most recent call last):
  File "/Users/brendan/.local/share/virtualenvs/core-bRkH1-KB/lib/python3.11/site-packages/langsmith/client.py", line 281, in _dumps_json_single
    return orjson.dumps(
           ^^^^^^^^^^^^^
TypeError: Dict key must a type serializable with OPT_NON_STR_KEYS

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/brendan/.local/share/virtualenvs/core-bRkH1-KB/lib/python3.11/site-packages/langsmith/client.py", line 5802, in _tracing_thread_handle_batch
    client.multipart_ingest_runs(create=create, update=update, pre_sampled=True)
  File "/Users/brendan/.local/share/virtualenvs/core-bRkH1-KB/lib/python3.11/site-packages/langsmith/client.py", line 1675, in multipart_ingest_runs
    valb = _dumps_json(value)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/brendan/.local/share/virtualenvs/core-bRkH1-KB/lib/python3.11/site-packages/langsmith/client.py", line 321, in _dumps_json
    return _dumps_json_single(obj, _serialize_json)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/brendan/.local/share/virtualenvs/core-bRkH1-KB/lib/python3.11/site-packages/langsmith/client.py", line 292, in _dumps_json_single
    result = json.dumps(
             ^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
          ^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
TypeError: keys must be str, int, float, bool or None, not PosixPath
```

### CODE REVIEW

**Comment by hinthornw** [comment]
Thanks for flagging and proposing a fix! Do you have an example test case you could add? :)

File: python/langsmith/_internal/_serde.py [hunk 1]
```diff
@@ -35,14 +31,8 @@ def _simple_default(obj):
...
             return obj.isoformat()
-        if isinstance(obj, uuid.UUID):
+        elif isinstance(obj, uuid.UUID):
             return str(obj)
-        if hasattr(obj, "model_dump") and callable(obj.model_dump):
-            return obj.model_dump()
-        elif hasattr(obj, "dict") and callable(obj.dict):
-            return obj.dict()
-        elif hasattr(obj, "_asdict") and callable(obj._asdict):
-            return obj._asdict()
```
Comment (brendanator) [inline python/langsmith/_internal/_serde.py:45]: removed these are they are dead code - namedtuple, and pydantic.BaseModel are handled in `_serialize_json`

File: python/langsmith/_internal/_serde.py [hunk 2]
```diff
@@ -88,17 +78,18 @@ def _serialize_json(obj: Any) -> Any:
             return list(obj)
 
         serialization_methods = [
-            ("model_dump", True),  # Pydantic V2 with non-serializable fields
-            ("dict", False),  # Pydantic V1 with non-serializable field
-            ("to_dict", False),  # dataclasses-json
+            (
+                "model_dump",
+                {"exclude_none": True, "mode": "json"},
```
Comment (brendanator) [inline python/langsmith/_internal/_serde.py:83]: mode="json" convert `pathlib.Path` dict keys to `str`

File: python/langsmith/_internal/_serde.py [hunk 3]
```diff
@@ -120,13 +111,11 @@ def _elide_surrogates(s: bytes) -> bytes:
     return result
 
 
-def _dumps_json_single(
-    obj: Any, default: Optional[Callable[[Any], Any]] = None
```
Comment (brendanator) [inline python/langsmith/_internal/_serde.py:124]: Removed the `default` parameter as `_dump_json_single` is only ever called with `_serialize_json` and falling back to `_simple_default` was confusing

**Comment by brendanator** [comment]
@hinthornw Added test cases, and then cleaned up my implementation

**Comment by brendanator** [comment]
@hinthornw Is there anything I need to do to get this in? Happy to make any follow up changes you want

**Comment by brendanator** [comment]
I've merged latest main and updated the tests

**Pull request merged**